bin/metrics-cgroup.rb

```diff
@@ -0,0 +1,149 @@
+#
+# PLATFORMS:
+#   Linux
+#
+# DEPENDENCIES:
+#   gem: sensu-plugin
+#
+# USAGE:
+#   # Collect all cgroup metrics, tested only on CentOS.
+#   metrics-cgroup.rb -g "/sys/fs/cgroup/**/**"
+#
+#   # Collect only systemd service metrics
+#   metrics-cgroup.rb -g "/sys/fs/cgroup/**/system.slice/*"
+#
+#   # Collect only user.slice only
+#   metrics-cgroup.rb -g "/sys/fs/cgroup/**/user.slice"
+#
+#   # Collect only mesos task metrics
+#   metrics-cgroup.rb -g "/sys/fs/cgroup/**/mesos/*"
+#
+#   # Collect only docker container metrics
+#   # (may not be meaningful if you already collect metrics from docker)
+#   metrics-cgroup.rb -g "/sys/fs/cgroup/**/docker/*"
+#
+#   # collect only system.slice, user.slice, docker and mesos and nothing else
+#   metrics-cgroup.rb -g "/sys/fs/cgroup/**/{system.slice,user.slice,docker,mesos}{,/*}"
+#
+#   # Instead of collecting breakdown of each docker instance, or
+#   # mesos task, or systemd service metric; this could let to
+#   # collect/compare overall docker, mesos, systemd and user recources
+#   # in total.
+#   metrics-cgroup.rb -g "/sys/fs/cgroup/**/{system.slice,user.slice,docker,mesos}{,/*}"
+#
+# LICENCE:
+#   Released under the same terms as Sensu (the MIT license); see LICENSE
+#   for details.
+#
+require 'sensu-plugin/metric/cli'
+require 'socket'
+
+class CgroupMetrics < Sensu::Plugin::Metric::CLI::Graphite
+  FILES_TO_SKIP = [
+    'cgroup.clone_children',
+    'cgroup.procs',
+    'notify_on_release',
+    'tasks'
+  ].freeze
+
+  option :scheme,
+         description: 'Metric naming scheme, text to prepend to queue_name.metric',
+         short: '-s SCHEME',
+         long: '--scheme SCHEME',
+         default: "#{Socket.gethostname}.cgroup"
+  option :glob,
+         description: 'Metric naming scheme, text to prepend to queue_name.metric',
+         short: '-g GLOB',
+         long: '--glob GLOB',
+         default: '/sys/fs/cgroup/**/**'
+  option :files,
+         description: 'Limit metric collection to this files, comma seperated list of files in cgroup directories. (default is to collect all files)',
+         short: '-f FILES',
+         long: '--files FILES',
+         default: ''
+  option :smart_paths,
+         description: 'Instead of using full cgroup path in metrics this enabling smarter metrics names',
+         short: '-p',
+         long: '--no-smart-paths',
+         boolean: true
+
+  def get_tag(cgroup_dir)
+    if config[:smart_paths]
+      "path.#{cgroup_dir.gsub(/[.\/]/, '_')}"
+    else
+      # use cgroup consumer specific tags
+      type, val = cgroup_dir.split(File::SEPARATOR)[-2..-1]
+      case type
+      when 'docker'
+        "docker_id.#{val}"
+      when 'mesos'
+        "mesos_container_id.#{val}"
+      when 'system.slice'
+        "systemd_unit.#{val.gsub(/[.\/]/, '_')}"
+      else
+        "path.#{cgroup_dir.gsub(/[.\/]/, '_')}"
+      end
+    end
+  end
+
+  def run
+    Dir.glob(config[:glob]).select { |d| File.directory?(d) }.each do |cgroup_dir|
```

Comment:
Warning is the primary reason in this case, other cases it matters more. Basically if you always expect something to be empty or !nil then when you see this you properly handle these as opposed to just jumping in which may or may not be an issue. If you have ever seen a nil class error in ruby this is how you better handle those situations. Rather than having a random generic nil class in ruby you have a more meaningful error/warning message.